p2p/profile_service.go

```diff
@@ -0,0 +1,218 @@
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/helpers"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/qri-io/qri/config"
+	"github.com/qri-io/qri/event"
+	"github.com/qri-io/qri/repo"
+	"github.com/qri-io/qri/repo/profile"
+)
+
+// ProfileProtocolID is the protocol id for the profile exchange service
+const ProfileProtocolID = "/qri/profile/0.1.0"
+
+// ProfileTimeout is the length of time we will wait for a response in a profile
+// exchange
+const ProfileTimeout = time.Minute * 2
+
+// ProfileService manages the profile exchange. This exchange should happen
+// whenever a node connects to a new peer
+type ProfileService struct {
+	Host    host.Host
+	Repo    repo.Repo
+	pub     event.Publisher
+	peersMu *sync.Mutex
+	peers   map[peer.ID]chan struct{}
+}
+
+// NewQriProfileService creates an profile exchange service and adds
+// the `ProfileHandler` to the host
+func NewQriProfileService(r repo.Repo, p event.Publisher) *ProfileService {
+	ps := &ProfileService{
+		// Host:    h,
+		Repo:    r,
+		pub:     p,
+		peersMu: &sync.Mutex{},
+		peers:   map[peer.ID]chan struct{}{},
+	}
+	// h.SetStreamHandler(ProfileProtocolID, ps.ProfileHandler)
+	return ps
+}
+
+// StartProfileService adds a host and a profile handler to the host
+// to the profile service
+func (ps *ProfileService) StartProfileService(h host.Host) {
+	ps.Host = h
+	h.SetStreamHandler(ProfileProtocolID, ps.ProfileHandler)
+}
+
+// ProfileHandler listens for profile requests
+// it sends it's node's profile on the given stream
+// whenever a request comes in
+func (ps *ProfileService) ProfileHandler(s network.Stream) {
+
+	var (
+		err error
+		pro *profile.Profile
+	)
+
+	ctx := context.Background()
+
+	p := s.Conn().RemotePeer()
+
+	defer func() {
+		helpers.FullClose(s)
+		if err != nil {
+			ps.pub.Publish(ctx, event.ETP2PProfileSendFailed, p)
+			return
+		}
+		ps.pub.Publish(ctx, event.ETP2PProfileSendSuccess, p)
+	}()
+
+	log.Debug("%s received a profile request from %s %s", ProfileProtocolID, p, s.Conn().RemoteMultiaddr())
+
+	if err := ps.pub.Publish(ctx, event.ETP2PProfileRequestReceived, p); err != nil {
+		log.Errorf("%s publishing event %s", ProfileProtocolID, err)
+	}
+
+	pro, err = ps.Repo.Profile()
+	if err != nil {
+		log.Errorf("%s error getting this node's profile: %s", ProfileProtocolID, err)
```

Comment:
I've been using recent PRs to transition logging within error handling to debug level. error-level messages create additional noise that confuses users.